connectors/uplynk/src/main/java/com/theoplayer/android/connector/uplynk/internal/UplynkAdIntegration.kt

```diff
@@ -38,30 +41,37 @@ internal class UplynkAdIntegration(
         val uplynkSource = source.sources.singleOrNull { it.ssai is UplynkSsaiDescription }
         val ssaiDescription = uplynkSource?.ssai as? UplynkSsaiDescription ?: return source
 
-        val response = uplynkDescriptionConverter
-            .buildPreplayUrl(ssaiDescription)
-            .let { uplynkApi.preplay(it) }
-            .also {
-                try {
-                    val response = it.parseExternalResponse()
-                    eventDispatcher.dispatchPreplayEvents(response)
-                    eventDispatcher.dispatchAdBreaksUpdatedEvents(response.ads)
-                    adScheduler = UplynkAdScheduler(response.ads.breaks, AdHandler(controller), eventDispatcher)
-                } catch (e: Exception) {
-                    eventDispatcher.dispatchPreplayFailure(e)
-                    controller.error(e)
-                }
-            }
+        val preplayUrl = uplynkDescriptionConverter.buildPreplayUrl(ssaiDescription)
+        val internalResponse = uplynkApi.preplay(preplayUrl)
+        val minimalResponse = internalResponse.parseMinimalResponse()
 
-        val internalResponse = response.parseMinimalResponse()
+        var newUplynkSource = uplynkSource.replaceSrc(minimalResponse.playURL)
+        minimalResponse.drm?.let { drm ->
+            val drmBuilder = DRMConfiguration.Builder()
+            drm.widevineLicenseURL?.let { drmBuilder.widevine(KeySystemConfiguration.Builder(it).build()) }
+            drm.playreadyLicenseURL?.let { drmBuilder.playready(KeySystemConfiguration.Builder(it).build()) }
+            drm.fairplayCertificateURL?.let { drmBuilder.fairplay(FairPlayKeySystemConfiguration.Builder("", it).build()) }
+            newUplynkSource = newUplynkSource.replaceDrm(drmBuilder.build())
+        }
```

Comment:
Nit-pick: I like to wrap my builder method calls in one big `apply` block. 🙂
```suggestion
            val drmBuilder = DRMConfiguration.Builder().apply {
                drm.widevineLicenseURL?.let { widevine(KeySystemConfiguration.Builder(it).build()) }
                drm.playreadyLicenseURL?.let { playready(KeySystemConfiguration.Builder(it).build()) }
                drm.fairplayCertificateURL?.let { fairplay(FairPlayKeySystemConfiguration.Builder("", it).build()) }
            }
            newUplynkSource = newUplynkSource.replaceDrm(drmBuilder.build())
        }
```